### PR TITLE
add mode to inclusion check

### PR DESCRIPTION
**Pull request by jordanjennings-mysten**
## Description 

Describe the changes or additions included in this PR.

## Test plan 

How did you test the new or updated feature?

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

File: external-crates/move/crates/move-binary-format/src/compatibility.rs
Comment (amnn) [inline external-crates/move/crates/move-binary-format/src/compatibility.rs:396]: Generally better to use an ordered set (like `BTreeSet`) for anything that's running in the execution layer. Will help to avoid any unintended forks.

Comment (amnn) [reply external-crates/move/crates/move-binary-format/src/compatibility.rs:396]: Also, is there a way for us to avoid allocating this set and removing elements from it in the case we only care that the sets are different? I.e. can we hand more responsibility to the mode to decide whether it wants to track the actual set of structs/enums/functions that were added, vs just knowing that they are different?

**Review by amnn — COMMENTED** [review]
Broadly this looks good to me, although I'd love to have @tzakian take a look as well. We will want to defer the creation of new struct/enum/function sets to the CLI mode impl (so the execution impl doesn't have to pay that cost), but that's the main comment here -- the structure looks good.

File: external-crates/move/crates/move-binary-format/src/compatibility.rs
Comment (jordanjennings-mysten) [reply external-crates/move/crates/move-binary-format/src/compatibility.rs:396]: since they are already sorted with btreemaps you could do a two pointer where each is simultaneously iterated over to check for the extra/missing, keeps runtime linear, should remove memory overhead.

File: external-crates/move/crates/move-binary-format/src/compatibility.rs
Comment (amnn) [inline external-crates/move/crates/move-binary-format/src/compatibility.rs:413]: Why not iterate over the key value pairs and return references to the values in your `Mark` struct to avoid the `unwrap`s?

File: external-crates/move/crates/move-binary-format/src/compatibility.rs
Comment (amnn) [inline external-crates/move/crates/move-binary-format/src/compatibility.rs:414]: You probably want to return the old enum here (the one that is missing) because in the error message, that will allow you to print out the full enum that needs to be added back.

File: external-crates/move/crates/move-binary-format/src/compatibility.rs
Comment (amnn) [inline external-crates/move/crates/move-binary-format/src/compatibility.rs:488]: This needs lots of unit tests.

File: external-crates/move/crates/move-binary-format/src/compatibility.rs
Comment (amnn) [inline external-crates/move/crates/move-binary-format/src/compatibility.rs:446]: I kind of want to suggest generalising `compare_identifier` so that it can be used here -- it should be fairly simple to do, -- it just needs to accept a comparator function to do the ordering with.

**Review by amnn — COMMENTED** [review]
LGTM modulo inline comments -- but it would still be good to get @tzakian to take a look and give the final stamp.

File: external-crates/move/crates/move-binary-format/src/compatibility.rs
Comment (jordanjennings-mysten) [reply external-crates/move/crates/move-binary-format/src/compatibility.rs]: going to try proptest, seems worth it for this area.

File: external-crates/move/crates/move-binary-format/src/unit_tests/compatibility_tests.rs
Comment (tzakian) [inline external-crates/move/crates/move-binary-format/src/unit_tests/compatibility_tests.rs:1217]: I'd use [`prop_assert`]( https://altsysrq.github.io/rustdoc/proptest/latest/proptest/macro.prop_assert.html) and `prop_assert_eq` here and elsewhere in this function as it will help proptest narrow counterexamples if errors are found.

File: external-crates/move/crates/move-binary-format/src/unit_tests/compatibility_tests.rs
Comment (tzakian) [inline external-crates/move/crates/move-binary-format/src/unit_tests/compatibility_tests.rs:1242]: If you use `BTreeSet`s here for `old_keys` and `new_keys` (and for `combined` etc) instead of vectors it may make the logic here simpler.

File: external-crates/move/crates/move-binary-format/src/compatibility.rs
Comment (tzakian) [inline external-crates/move/crates/move-binary-format/src/compatibility.rs:425]: This can/should be updated to use the values returned from the `compare_ord_iters`

**Comment by jordanjennings-mysten** [comment]
also don't need the peeks so I dropped that and dropped the unwraps in compare_ord_iters.

**Review by tzakian — CHANGES_REQUESTED** [review]
Chatted offline and there's one or two things that should probably be changed before landing so requesting changes for that.

**Review by tzakian — APPROVED** [review]
LGTM! For real this time :)

**Pull request merged**
